parlai/scripts/data_stats.py

```diff
@@ -130,54 +119,52 @@ def keep_token(t):
-                    counts['tokens_in_' + itype] / counts['utterances_in_' + itype]
-                )
-                counts['avg_utterance_length_in_' + 'both'] = (
-                    counts['tokens_in_' + 'both'] / counts['utterances_in_' + 'both']
-                )
+                retxt = [t for t in tokens if keep_token(t)]
+                counts[f'{itype}/tokens'] += len(retxt)
+                counts['both/tokens'] += len(retxt)
+                counts[f'{itype}/utterances'] += 1
+                counts['both/utterances'] += 1
+                counts[f'{itype}/avg_utterance_length'] += AverageMetric(len(retxt), 1)
+                counts[f'both/avg_utterance_length'] += AverageMetric(len(retxt), 1)
                 for t in retxt:
-                    if t not in counts['token_dict_' + itype]:
-                        counts['unique_tokens_in_' + itype] += 1
-                        counts['token_dict_' + itype][t] = True
-                    if t not in counts['token_dict_' + 'both']:
-                        counts['unique_tokens_in_' + 'both'] += 1
-                        counts['token_dict_' + 'both'][t] = True
+                    if t not in counts[f'{itype}/token_dict']:
+                        counts[f'{itype}/unique_tokens'] += 1
+                        counts[f'{itype}/token_dict'][t] = True
+                    if t not in counts['both/token_dict']:
+                        counts['both/unique_tokens'] += 1
+                        counts['both/token_dict'][t] = True
                 retxt = ' '.join(retxt)
-                if retxt not in counts['utterance_dict_' + itype]:
-                    counts['unique_utterances_in_' + itype] += 1
-                    counts['utterance_dict_' + itype][retxt] = True
-                if retxt not in counts['utterance_dict_' + 'both']:
-                    counts['unique_utterances_in_' + 'both'] += 1
-                    counts['utterance_dict_' + 'both'][retxt] = True
+                if retxt not in counts[f'{itype}/utterance_dict']:
+                    counts[f'{itype}/unique_utterances'] += 1
+                    counts[f'{itype}/utterance_dict'][retxt] = True
+                if retxt not in counts['both/utterance_dict']:
+                    counts['both/unique_utterances'] += 1
+                    counts['both/utterance_dict'][retxt] = True
 
         if log_time.time() > log_every_n_secs:
-            text, log = report(world, counts, log_time)
-            if print_parser:
-                logging.info(text)
+            report = _report(world, counts)
+            cnt = report.pop('exs')
+            text, log = log_time.log(cnt, world.num_examples(), report)
+            logging.info(text)
 
     try:
         # print dataset size if available
         logging.info(
             f'loaded {world.num_episodes()} episodes with a total '
             f'of {world.num_examples()} examples'
         )
-    except Exception:
+    except AttributeError:
```

Comment:
when does a world not have these attributes?

Reply:
some worlds don't! streaming ones.